### PR TITLE
Allow br and hr without closing />

### LINKED ISSUE #41: Allow <br> and <hr> without closing />
The validator should throw nothing (no error etc.) when `<br>, <hr>` instead of `<hr />` is used: https://travis-ci.org/milescellar/phpbb-language-fr/jobs/296454250

### PR DESCRIPTION
**Pull request by Crizz0**
Fixes #41 

@milescellar Can you test this with your language package?

### CODE REVIEW

**Comment by milescellar** [comment]
@Crizz0 Tested. Some strings passed, but some don't (when there are multiple `<br>` tags, that won't pass).

Example: https://travis-ci.org/milescellar/phpbb-language-fr/jobs/297614141

**Comment by Crizz0** [comment]
Thanks for testing. 

Please try again. :)

**Comment by milescellar** [comment]
@Crizz0 It looks good now! Example: https://travis-ci.org/milescellar/phpbb-language-fr/jobs/297614141

**Comment by milescellar** [comment]
However, I just tried to replace all old XHTML closing tags with HTML5 ones for consistency code, but the validator seems to see additional tags when these changes are not applied to the British English language pack.

Commit: https://github.com/milescellar/phpbb-language-fr/commit/a9fe9f6e0ea0dd3a7d48762562c33bd2449a749b
Failing build: https://travis-ci.org/milescellar/phpbb-language-fr/jobs/297680400

Also, there is still a missing tag error with the `` tag, when I replace `</img>` with `</img>` (which is HTML5 valid).

No more issue with other tags. 😉

**Comment by Crizz0** [comment]
Yes, it is. I will take look.

Both versions should be allowed and don't lead to a test failure.

**Comment by Crizz0** [comment]
@milescellar Are you sure, that you use the last version of this PR? I got only notices, if I have converted the   `<br />` in 4 cases to `<br>`, but no FATAL errors for them. Although my english-language packages has still its `<br />` there in the same place.

**Comment by milescellar** [comment]
Yes I'm sure. I changed all the `<br />` tags to `<br>` when it's possible on the language files and I'm having these fatal errors. It's because the validator just compare my language pack to the British English one and see that there are different tags (because it don't know that the `<br>` and `<br />` tags have the same function).

**Comment by Crizz0** [comment]
The thing is, I cannot reproduce this, yet. I did just not have this FATAL error their with my tests with the german language and the 3.2 en/-original one. I'll try yours.

**Comment by milescellar** [comment]
I just tried again with the missing code from your commit, sorry for that.

Most fatal errors have been fixed, but some are still present in 9 strings:
https://travis-ci.org/milescellar/phpbb-language-fr/jobs/298454766

_(Note that I just replaced all `<br />` tags to `<br>`, but I didn't replace  to  where there was also a fatal error)_

File: src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php [hunk 1]
```diff
@@ -675,12 +675,18 @@ public function validateHtml($file, $key, $sourceString, $originString)
...
+				'hr',
+			);
+
 			$tag = (strpos($possibleHtml, ' ') !== false) ? substr($possibleHtml, 1, strpos($possibleHtml, ' ') - 1) : substr($possibleHtml, 1, strpos($possibleHtml, '>') - 1);
 			$tag = ($openingTag) ? $tag : substr($tag, 1);
 
 			if ($openingTag)
 			{
-				if (in_array($tag, $openTags))
+				if (in_array($tag, $openTags) && !in_array($tag, $allowedMissingClosingTag))
```
Comment (nickvergessen) [inline src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php]: The addtion from here should be on line old691/new697:
```php
else if (substr($possibleHtml, -3) !== ' />')
```

File: src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php [hunk 2]
```diff
@@ -749,7 +755,7 @@ public function validateHtml($file, $key, $sourceString, $originString)
 			}
 		}
 
-		if (!empty($openTags) && !$failedUnclosed)
+		if (!empty($openTags) && !$failedUnclosed && !in_array($openTags[0], $allowedMissingClosingTag))
```
Comment (nickvergessen) [inline src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php]: this should be reverted

**Comment by Crizz0** [comment]
Works, but for  is complains still for additional html with Line 755. Maybe an exception would be okay for this if-sentence.

**Comment by milescellar** [comment]
I confirm, the new patch works well but there is still a fatal error for the `` tag.
https://travis-ci.org/milescellar/phpbb-language-fr/jobs/298471882

File: src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php [hunk 3]
```diff
@@ -688,7 +695,7 @@ public function validateHtml($file, $key, $sourceString, $originString)
 					}
 					$failedUnclosed = true;
 				}
-				else if (substr($possibleHtml, -3) !== ' />')
+				else if ((substr($possibleHtml, -3) !== ' />') && !in_array($tag, $allowedMissingClosingTag))
```
Comment (nickvergessen) [inline src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php]: the brackets you added are not necessary

**Comment by milescellar** [comment]
The latest commit (251e954) is causing a build failure.

**Comment by Crizz0** [comment]
It looks like, this brackets are necessary. Because without it throws:
`PHP Parse error:  syntax error, unexpected '&&' (T_BOOLEAN_AND) in /home/travis/build/phpbb/phpbb-translation-validator/src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php on line 698`

File: src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php [hunk 4]
```diff
@@ -695,7 +695,7 @@ public function validateHtml($file, $key, $sourceString, $originString)
 					}
 					$failedUnclosed = true;
 				}
-				else if ((substr($possibleHtml, -3) !== ' />') && !in_array($tag, $allowedMissingClosingTag))
+				else if (substr($possibleHtml, -3) !== ' />') && !in_array($tag, $allowedMissingClosingTag)
```
Comment (nickvergessen) [inline src/Phpbb/TranslationValidator/Validator/LangKeyValidator.php]: Replace:
```php
else if (substr($possibleHtml, -3) !== ' />') && !in_array($tag, $allowedMissingClosingTag)
```
with:
```php
else if (substr($possibleHtml, -3) !== ' />' && !in_array($tag, $allowedMissingClosingTag))
```

**Comment by Crizz0** [comment]
@milescellar Any other problems with this patch?

**Comment by milescellar** [comment]
There is still a fatal error concerning the `` tag.

**Comment by Crizz0** [comment]
Yes, but I seems like, the validator doesn't recognizes the img without / as additional HTML, no matter who is missing the /> (en or your/any language to validate).

Maybe I'll push this to 1.4.1 and get 1.4.0 online for 3.2.2 translations, first.

**Comment by milescellar** [comment]
Yeap, or the other way to fix it is to remove that slash in the language file itself (to keep XHTML tag specs is not a good idea). But I agree with you, let's release 1.4.0 for 3.2.2 translations, it's not a big issue. 🙂

**Pull request merged**